# LAGOON-MCL: Citations

## [Nextflow](https://www.nextflow.io/)

> P. Di Tommaso, M. Chatzou, E. W. Floden, P. P. Barja, E. Palumbo, et C. Notredame, « Nextflow enables reproducible computational workflows », Nat Biotechnol, vol. 35, no 4, p. 316‑319, avr. 2017, doi: [10.1038/nbt.3820](https://doi.org/10.1038/nbt.3820).

## Pipeline tools

* [**Diamond**](https://github.com/bbuchfink/diamond)

> B. Buchfink, K. Reuter, et H.-G. Drost, « Sensitive protein alignments at tree-of-life scale using DIAMOND », Nat Methods, vol. 18, no 4, Art. no 4, avr. 2021, doi: [10.1038/s41592-021-01101-x](https://doi.org/10.1038/s41592-021-01101-x).

* [**Markov CLustering algorithm**](https://micans.org/mcl/)

> A. J. Enright, S. Van Dongen, et C. A. Ouzounis, « An efficient algorithm for large-scale detection of protein families », Nucleic Acids Research, vol. 30, no 7, p. 1575‑1584, avr. 2002, doi: [10.1093/nar/30.7.1575](https://doi.org/10.1093/nar/30.7.1575).

> S. van Dongen et C. Abreu-Goodger, « Using MCL to Extract Clusters from Networks », in Bacterial Molecular Networks: Methods and Protocols, J. van Helden, A. Toussaint, et D. Thieffry, Éd., New York, NY: Springer, 2012, p. 281‑295. doi: [10.1007/978-1-61779-361-5_15](https://doi.org/10.1007/978-1-61779-361-5_15).

> S. Van Dongen, « Graph Clustering Via a Discrete Uncoupling Process », SIAM J. Matrix Anal. Appl., vol. 30, no 1, p. 121‑141, janv. 2008, doi: [10.1137/040608635](https://doi.org/10.1137/040608635).

* [**SeqKit2**](https://bioinf.shenwei.me/seqkit/)

> W. Shen, B. Sipos, et L. Zhao, « SeqKit2: A Swiss army knife for sequence and alignment processing », iMeta, vol. 3, no 3, p. e191, 2024, doi: [10.1002/imt2.191](https://doi.org/10.1002/imt2.191).

* [**MMseqs2**](https://github.com/soedinglab/MMseqs2)

> M. Steinegger et J. Söding, « MMseqs2 enables sensitive protein sequence searching for the analysis of massive data sets », Nat Biotechnol, vol. 35, no 11, Art. no 11, nov. 2017, doi: [10.1038/nbt.3988](https://doi.org/10.1038/nbt.3988).

## Pipeline database

* [**Pfam database**](http://pfam.xfam.org/)

> Mistry, J., Chuguransky, S., Williams, L., Qureshi, M., Salazar, G.A., Sonnhammer, E.L.L., Tosatto, S.C.E., Paladin, L., Raj, S., Richardson, L.J., Finn, R.D., Bateman, A., 2021. Pfam: The protein families database in 2021. Nucleic Acids Research 49, D412–D419. https://doi.org/10.1093/nar/gkaa913

## Software packaging / containerisation tools

* [**Singularity**](https://sylabs.io/singularity/)

> G. M. Kurtzer, V. Sochat, et M. W. Bauer, « Singularity: Scientific containers for mobility of compute », PLOS ONE, vol. 12, no 5, p. e0177459, mai 2017, doi: [10.1371/journal.pone.0177459](https://doi.org/10.1371/journal.pone.0177459).

* [**BioContainers**](https://biocontainers.pro/)

> F. da Veiga Leprevost et al., « BioContainers: an open-source and community-driven framework for software standardization », Bioinformatics, vol. 33, no 16, p. 2580‑2582, août 2017, doi: 10.1093/bioinformatics/btx192.

* [**Docker**](https://www.docker.com/)

> D. Merkel, « Docker: lightweight Linux containers for consistent development and deployment », Linux J., vol. 2014, no 239, p. 2:2, mars 2014.
