# LAGOON-MCL: Citations

## [Nextflow](https://www.nextflow.io/)

> Di Tommaso, P., Chatzou, M., Floden, E. W., Barja, P. P., Palumbo, E., & Notredame, C. (2017). Nextflow enables reproducible computational workflows. Nature Biotechnology, 35(4), 316‑319. [https://doi.org/10.1038/nbt.3820](https://doi.org/10.1038/nbt.3820)

## Pipeline tools

* [**Diamond**](https://github.com/bbuchfink/diamond)

> Buchfink, B., Reuter, K., & Drost, H.-G. (2021). Sensitive protein alignments at tree-of-life scale using DIAMOND. Nature Methods, 18(4), Article 4. [https://doi.org/10.1038/s41592-021-01101-x](https://doi.org/10.1038/s41592-021-01101-x)

* [**Markov CLustering algorithm**](https://micans.org/mcl/)

> Enright, A. J., Van Dongen, S., & Ouzounis, C. A. (2002). An efficient algorithm for large-scale detection of protein families. Nucleic Acids Research, 30(7), 1575‑1584.

> Van Dongen, S. (2008). Graph Clustering Via a Discrete Uncoupling Process. SIAM Journal on Matrix Analysis and Applications, 30(1), 121‑141. [https://doi.org/10.1137/040608635](https://doi.org/10.1137/040608635)

> van Dongen, S., & Abreu-Goodger, C. (2012). Using MCL to Extract Clusters from Networks. In J. van Helden, A. Toussaint, & D. Thieffry (Éds.), Bacterial Molecular Networks : Methods and Protocols (p. 281‑295). Springer. [https://doi.org/10.1007/978-1-61779-361-5_15](https://doi.org/10.1007/978-1-61779-361-5_15)

* [**SeqKit2**](https://bioinf.shenwei.me/seqkit/)

> Shen, W., Sipos, B., & Zhao, L. (2024). SeqKit2 : A Swiss army knife for sequence and alignment processing. iMeta, 3(3), e191. [https://doi.org/10.1002/imt2.191](https://doi.org/10.1002/imt2.191)

* [**MMseqs2**](https://github.com/soedinglab/MMseqs2)

> Steinegger, M., & Söding, J. (2017). MMseqs2 enables sensitive protein sequence searching for the analysis of massive data sets. Nature Biotechnology, 35(11), Article 11. [https://doi.org/10.1038/nbt.3988](https://doi.org/10.1038/nbt.3988)

## Pipeline databases

* [**Pfam database**](http://pfam.xfam.org/)

> Mistry, J., Chuguransky, S., Williams, L., Qureshi, M., Salazar, G. A., Sonnhammer, E. L. L., Tosatto, S. C. E., Paladin, L., Raj, S., Richardson, L. J., Finn, R. D., & Bateman, A. (2021). Pfam : The protein families database in 2021. Nucleic Acids Research, 49(D1), D412‑D419. [https://doi.org/10.1093/nar/gkaa913](https://doi.org/10.1093/nar/gkaa913)

* [**AlphaFold clusters database**]

> Barrio-Hernandez, I., Yeo, J., Jänes, J., Mirdita, M., Gilchrist, C. L. M., Wein, T., Varadi, M., Velankar, S., Beltrao, P., & Steinegger, M. (2023). Clustering predicted structures at the scale of the known protein universe. Nature, 622(7983), 637‑645. [https://doi.org/10.1038/s41586-023-06510-w](https://doi.org/10.1038/s41586-023-06510-w)

> Varadi, M., Bertoni, D., Magana, P., Paramval, U., Pidruchna, I., Radhakrishnan, M., Tsenkov, M., Nair, S., Mirdita, M., Yeo, J., Kovalevskiy, O., Tunyasuvunakool, K., Laydon, A., Žídek, A., Tomlinson, H., Hariharan, D., Abrahamson, J., Green, T., Jumper, J., … Velankar, S. (2024). AlphaFold Protein Structure Database in 2024 : Providing structure coverage for over 214 million protein sequences. Nucleic Acids Research, 52(D1), D368‑D375. [https://doi.org/10.1093/nar/gkad1011](https://doi.org/10.1093/nar/gkad1011)

> Jumper, J., Evans, R., Pritzel, A., Green, T., Figurnov, M., Ronneberger, O., Tunyasuvunakool, K., Bates, R., Žídek, A., Potapenko, A., Bridgland, A., Meyer, C., Kohl, S. A. A., Ballard, A. J., Cowie, A., Romera-Paredes, B., Nikolov, S., Jain, R., Adler, J., … Hassabis, D. (2021). Highly accurate protein structure prediction with AlphaFold. Nature, 596(7873), Article 7873. [https://doi.org/10.1038/s41586-021-03819-2](https://doi.org/10.1038/s41586-021-03819-2)

## Software packaging / containerisation tools

* [**Singularity**](https://sylabs.io/singularity/)

> Kurtzer, G. M., Sochat, V., & Bauer, M. W. (2017). Singularity : Scientific containers for mobility of compute. PLOS ONE, 12(5), e0177459. [https://doi.org/10.1371/journal.pone.0177459](https://doi.org/10.1371/journal.pone.0177459)

* [**BioContainers**](https://biocontainers.pro/)

> da Veiga Leprevost, F., Grüning, B. A., Alves Aflitos, S., Röst, H. L., Uszkoreit, J., Barsnes, H., Vaudel, M., Moreno, P., Gatto, L., Weber, J., Bai, M., Jimenez, R. C., Sachsenberg, T., Pfeuffer, J., Vera Alvarez, R., Griss, J., Nesvizhskii, A. I., & Perez-Riverol, Y. (2017). BioContainers : An open-source and community-driven framework for software standardization. Bioinformatics (Oxford, England), 33(16), 2580‑2582. [https://doi.org/10.1093/bioinformatics/btx192](https://doi.org/10.1093/bioinformatics/btx192)

* [**Docker**](https://www.docker.com/)

> Merkel, D. (2014). Docker : Lightweight Linux containers for consistent development and deployment. Linux J., 2014(239), 2:2.